pandas/core/frame.py

```diff
@@ -9241,7 +9241,12 @@ def mode(
         def f(s):
             return s.mode(dropna=dropna)
 
-        return data.apply(f, axis=axis)
+        data = data.apply(f, axis=axis)
+        # Ensure index is type stable (should always use int index)
+        if data.empty:
+            data.index = Index([], dtype="int64")
+
+        return data
```

Comment:
Empty result case doesn't hit `algorithms.mode`, so those changes aren't enough to ensure correct `dtype`